services/apps/alcs/src/alcs/notice-of-intent/notice-of-intent-tag/notice-of-intent-tag.service.ts

```diff
@@ -0,0 +1,84 @@
+import { Injectable, Logger } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Tag } from '../../tag/tag.entity';
+import { NoticeOfIntent } from '../notice-of-intent.entity';
+import { Repository } from 'typeorm';
+import { ServiceNotFoundException, ServiceValidationException } from '@app/common/exceptions/base.exception';
+
+@Injectable()
+export class NoticeOfIntentTagService {
+  private logger = new Logger(NoticeOfIntentTagService.name);
+
+  constructor(
+    @InjectRepository(Tag) private tagRepository: Repository<Tag>,
+    @InjectRepository(NoticeOfIntent) private noiRepository: Repository<NoticeOfIntent>,
+  ) {}
+
+  async addTagToNoticeOfIntent(fileNumber: string, tagName: string) {
+    const noi = await this.noiRepository.findOne({
+      where: { fileNumber: fileNumber },
+      relations: ['tags'],
+    });
+    if (!noi) {
+      throw new ServiceNotFoundException(`Notice of Intent not found with number ${fileNumber}`);
+    }
+
+    const tag = await this.tagRepository.findOne({ where: { name: tagName } });
+    if (!tag) {
+      throw new ServiceNotFoundException(`Tag not found with name ${tagName}`);
+    }
+
+    if (!noi.tags) {
+      noi.tags = [];
+    }
+
+    const tagExists = noi.tags.some((t) => t.uuid === tag.uuid);
+    console.log(tagExists);
+    if (tagExists) {
+      throw new ServiceValidationException(`Tag ${tagName} already exists`);
+    }
+
+    noi.tags.push(tag);
+    return this.noiRepository.save(noi);
+  }
+
+  async removeTagFromNoticeOfIntent(fileNumber: string, tagName: string) {
+    const noi = await this.noiRepository.findOne({
+      where: { fileNumber: fileNumber },
+      relations: ['tags'],
+    });
+    if (!noi) {
+      throw new ServiceNotFoundException(`Notice of Intent not found with number ${fileNumber}`);
+    }
+
+    const tag = await this.tagRepository.findOne({ where: { name: tagName } });
+    if (!tag) {
+      throw new ServiceNotFoundException(`Tag not found with name ${tagName}`);
+    }
+
+    if (!noi.tags) {
+      noi.tags = [];
+    }
+
+    const tagExists = noi.tags.some((t) => t.uuid === tag.uuid);
+    if (!tagExists) {
+      throw new ServiceValidationException(`Tag ${tagName} does not exist`);
+    }
+
+    noi.tags = noi.tags.filter((t) => t.uuid !== tag.uuid);
+    return this.noiRepository.save(noi);
+  }
+
+  async getNoticeOfIntentTags(fileNumber: string) {
+    const noi = await this.noiRepository.findOne({
+      where: { fileNumber: fileNumber },
+      relations: ['tags'],
+    });
+    if (!noi) {
+      throw new ServiceNotFoundException(`Notice of Intent not found with number ${fileNumber}`);
+    }
+    return noi.tags && noi.tags.length > 0
+      ? noi.tags.sort((a, b) => a.auditCreatedAt.getTime() - b.auditCreatedAt.getTime())
```

Comment:
Same comment I made on #1945: I generally feel sorting should be done on database.